File: flow/models/Service.rb
# -------------------------------------------------------------------------- #
# Copyright 2010-2013, C12G Labs S.L.                                        #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

module OpenNebula
    class Service < DocumentJSON

        DOCUMENT_TYPE = 100

        STATE = {
            'PENDING'            => 0,
            'DEPLOYING'          => 1,
            'RUNNING'            => 2,
            'UNDEPLOYING'        => 3,
            'FAILED'             => 4,
            'UNKNOWN'            => 5,
            'DONE'               => 6,
            'FAILED_UNDEPLOYING' => 7,
            'FAILED_DEPLOYING'   => 8
        }

        STATE_STR = [
            'PENDING',
            'DEPLOYING',
            'RUNNING',
            'UNDEPLOYING',
            'FAILED',
            'UNKNOWN',
            'DONE',
            'FAILED_UNDEPLOYING',
            'FAILED_DEPLOYING'
        ]

        LOG_COMP = "SER"

        # Returns the service state
        # @return [Integer] the service state
        def state
            return @body['state'].to_i
        end

        # Returns the service strategy
        # @return [String] the service strategy
        def strategy
            return @body['deployment']
        end

        # Returns the string representation of the service state
        # @return the state string
        def state_str
            return STATE_STR[state]
        end

        # Sets a new state
        # @param [Integer] the new state
        # @return [true, false] true if the value was changed
        def set_state(state)
            if state < 0 || state > 8
                return false
            end

            @body['state'] = state

            msg = "New state: #{STATE_STR[state]}"
            Log.info LOG_COMP, msg, self.id()
            self.log_info(msg)

            return true
        end

        # Returns the owner username
        # @return [String] the service's owner username
        def owner_name()
            return self['UNAME']
        end

        # Replaces this object's client with a new one
        # @param [OpenNebula::Client] owner_client the new client
        def replace_client(owner_client)
            @client = owner_client
        end

        # Returns all the node Roles
        # @return [Hash<String,Role>] all the node Roles
        def get_roles
            return @roles
        end

        # Returns true if all the nodes are correctly deployed
        # @return [true, false] true if all the nodes are correctly deployed
        def all_roles_running?()
            @roles.each { |name, role|
                if role.state != Role::STATE['RUNNING']
                    return false
                end
            }

            return true
        end

        # Returns true if all the nodes are in done state
        # @return [true, false] true if all the nodes are correctly deployed
        def all_roles_done?()
            @roles.each { |name, role|
                if role.state != Role::STATE['DONE']
                    return false
                end
            }

            return true
        end

        # Returns true if any of the roles is in failed state
        # @return [true, false] true if any of the roles is in failed state
        def any_role_failed?()
            failed_states = [
                Role::STATE['FAILED'], 
                Role::STATE['FAILED_DEPLOYING'], 
                Role::STATE['FAILED_UNDEPLOYING']]
                
            @roles.each { |name, role|
                if failed_states.include?(role.state)
                    return true
                end
            }

            return false
        end

        # Create a new service based on the template provided
        # @param [String] template_json
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(template_json)
            template = JSON.parse(template_json)
            template['state'] = STATE['PENDING']

            super(template.to_json, template['name'])
        end

        # Shutdown the service. This action is called when user wants to shutdwon
        #   the Service or it is in FAILED_UNDEPLOYING and the user has fixed the 
        #   problem
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def shutdown
            if [Service::STATE['RUNNING'], 
                    Service::STATE['FAILED_UNDEPLOYING'],
                    Service::STATE['UNKNOWN']].include?(self.state)
                self.set_state(Service::STATE['UNDEPLOYING'])
                return self.update
            else
                return OpenNebula::Error.new("Action shutdown: Wrong state" \
                    " #{self.state_str()}")
            end
        end

        # Deploy the service. This action is called when the Service is
        #   in FAILED_DEPLOYING and the user has fixed the problem and wants
        #   the deployment to continue
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def deploy
            if [Service::STATE['FAILED_DEPLOYING']].include?(self.state)
                self.set_state(Service::STATE['DEPLOYING'])
                return self.update
            else
                return OpenNebula::Error.new("Action deploy: Wrong state" \
                    " #{self.state_str()}")
            end
        end

        # Delete the service. All the VMs are also deleted from OpenNebula.
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def delete
            @roles.each { |name, role|
                role.delete()
            }

            return super()
        end

        # Retrieves the information of the Service and all its Nodes.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def info
            rc = super
            if OpenNebula.is_error?(rc)
                return rc
            end

            @roles = {}

            if @body['roles']
                @body['roles'].each { |elem|
                    elem['state'] ||= Role::STATE['PENDING']
                    role = Role.new(elem, self)
                    @roles[role.name] = role
                }
            end

            return nil
        end

        # Add an info message in the service information that will be stored
        #   in OpenNebula
        # @param [String] message
        def log_info(message)
            add_log(Logger::INFO, message)
        end

        # Add an error message in the service information that will be stored
        #   in OpenNebula
        # @param [String] message
        def log_error(message)
            add_log(Logger::ERROR, message)
        end

        # Retrieve the service client
        def client
            @client
        end

        # Changes the owner/group
        #
        # @param [Integer] uid the new owner id. Set to -1 to leave the current one
        # @param [Integer] gid the new group id. Set to -1 to leave the current one
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chown(uid, gid)
            old_uid = self['UID'].to_i
            old_gid = self['GID'].to_i

            rc = super(uid, gid)

            if OpenNebula.is_error?(rc)
                return rc
            end

            @roles.each { |name, role|
                rc = role.chown(uid, gid)

                break if rc[0] == false
            }

            if rc[0] == false
                self.log_error("Chown operation failed, will try to rollback all VMs to the old user and group")
                update()

                super(old_uid, old_gid)

                @roles.each { |name, role|
                    role.chown(old_uid, old_gid)
                }

                return OpenNebula::Error.new(rc[1])
            end

            return nil
        end

        private

        # @param [Logger::Severity] severity
        # @param [String] message
        def add_log(severity, message)
            severity_str = Logger::SEV_LABEL[severity][0..0]

            @body['log'] ||= Array.new
            @body['log'] << {
                :timestamp => Time.now.to_i,
                :severity  => severity_str,
                :message   => message
            }
        end
    end
end

File: src/sunstone/public/app/opennebula/service.js
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */

define(function(require) {
  var OpenNebulaAction = require('./action');
  var OpenNebulaHelper = require("./helper");
  var OpenNebulaError  = require("./error");
  var Locale = require('utils/locale');
  var Notifier = require('utils/notifier');

  var RESOURCE = "DOCUMENT";
  var PATH = 'service';
  var CACHE_NAME = 'SERVICE'

  var STATES_STR = [
    Locale.tr("PENDING"),
    Locale.tr("DEPLOYING"),
    Locale.tr("RUNNING"),
    Locale.tr("UNDEPLOYING"),
    Locale.tr("WARNING"),
    Locale.tr("DONE"),
    Locale.tr("FAILED_UNDEPLOYING"),
    Locale.tr("FAILED_DEPLOYING"),
    Locale.tr("SCALING"),
    Locale.tr("FAILED_SCALING"),
    Locale.tr("COOLDOWN"),
    Locale.tr("DEPLOYING_NETS"),
    Locale.tr("UNDEPLOYING_NETS"),
    Locale.tr("FAILED_DEPLOYING_NETS"),
    Locale.tr("FAILED_UNDEPLOYING_NETS")
  ];

  var STATES = {
    PENDING                 : 0,
    DEPLOYING               : 1,
    RUNNING                 : 2,
    UNDEPLOYING             : 3,
    WARNING                 : 4,
    DONE                    : 5,
    FAILED_UNDEPLOYING      : 6,
    FAILED_DEPLOYING        : 7,
    SCALING                 : 8,
    FAILED_SCALING          : 9,
    COOLDOWN                : 10,
    DEPLOYING_NETS          : 11,
    UNDEPLOYING_NETS        : 12,
    FAILED_DEPLOYING_NETS   : 13,
    FAILED_UNDEPLOYING_NETS : 14
  };

  var Service = {
    "resource": RESOURCE,
    "del": function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.del(params, RESOURCE, PATH);
    },
    "list" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.list(params, RESOURCE, PATH)
    },
    "show" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.show(params, RESOURCE, false, PATH)
    },
    "chown" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.chown(params, RESOURCE, PATH);
    },
    "chgrp" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.chgrp(params, RESOURCE, PATH);
    },
    "chmod" : function(params) {
      params.cache_name = CACHE_NAME;
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "chmod", action_obj, PATH);
    },
    "shutdown" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.simple_action(params, RESOURCE, "shutdown", null, PATH);
    },
    "recover" : function(params) {
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.simple_action(params, RESOURCE, "recover", null, PATH);
    },
    "recover_delete" : function(params) {
      var action_obj = {"delete": true};
      params.cache_name = CACHE_NAME;
      OpenNebulaAction.simple_action(params, RESOURCE, "recover", action_obj, PATH);
    },
    "rename" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "rename", action_obj, PATH);
    },
    "append": function(params) {
      params.cache_name = CACHE_NAME;

      var action_obj = {};

      try {
        JSON.parse(params.data.extra_param);
        action_obj["template"] = params.data.extra_param;
      }
      catch(err) {
        action_obj["template_raw"] = params.data.extra_param;
      }

      action_obj["append"] = true;

      var callback = params.success;
      var callbackError = params.error;
      var id = params.data.id;
      var request = OpenNebulaHelper.request(RESOURCE, "update", [id, action_obj]);

      var reqPath = PATH ? PATH : RESOURCE.toLowerCase();
      $.ajax({
        url: reqPath + "/" + id,
        type: "PUT",
        contentType: "application/json; charset=utf-8",
        data: JSON.stringify(action_obj),
        success: function(response) {
          return callback ? callback(request, response) : null;
        },
        error: function(response) {
          return callbackError ?
              callbackError(request, OpenNebulaError(response)) : null;
        }
      });
    },
    "update": function(params){
      params.cache_name = CACHE_NAME;

      var json_aux = JSON.parse(params.data.extra_param);
      var action_obj = JSON.stringify(json_aux);

      var callback = params.success;
      var callbackError = params.error;
      var id = params.data.id;
      var request = OpenNebulaHelper.request(RESOURCE, "update", [id, action_obj]);

      var reqPath = PATH ? PATH : RESOURCE.toLowerCase();

      $.ajax({
        url: reqPath + "/" + id,
        type: "PUT",
        contentType: "application/json; charset=utf-8",
        data: action_obj,
        success: function(response) {
          return callback ? callback(request, response) : null;
        },
        error: function(response) {
          return callbackError ?
              callbackError(request, OpenNebulaError(response)) : null;
        }
      });
    },
    "purgeDone": function(){
      var reqPath = PATH ? PATH : RESOURCE.toLowerCase();
      $.ajax({
        url: reqPath + '/purge',
        type: "POST",
        data: {
          csrftoken: csrftoken
        },
        success: function() {
          Notifier.notifyCustom(Locale.tr("Purge Services"),Locale.tr("Services on DONE state has been deleted from database"));
          return null;
        },
        error: Notifier.onError
      });
    },
    "stateStr" : function(stateId) {
      return STATES_STR[stateId];
    },
    "STATES": STATES,
    "filterDoneServices": _filterDoneServices,
    "getService": _getService,
    "getName": function(id){
      return OpenNebulaAction.getName(id, CACHE_NAME);
    }
  }

  function _promiseGetService({ id, success, async = true } = {}) {
    return $.ajax({
      url: 'service/' + id,
      type: 'GET',
      success: success,
      async: async
    });
  }

  function _getServiceById({ services, id } = {}) {
    return Array.isArray(services) &&
      services.find(function(service) {
        return service && service[RESOURCE] && service[RESOURCE].ID === id
      });
  }

  function _getService(id) {
    if (!id) return undefined;

    var service = undefined;
    var cache = OpenNebulaAction.cache(CACHE_NAME);

    if (cache && cache.data) {
      service = _getServiceById({ services: cache.data, id: id })
    }

    if (!service || service.length === 0) {
      _promiseGetService({
        id: id,
        async: false,
        success: function(res) { service = res }
      })
    }

    return service ? service[RESOURCE] : undefined
  }

  function _filterDoneServices(services) {
    return $.grep(services, function(service) {
      return service.DOCUMENT.TEMPLATE.BODY.state !== STATES.DONE;
    });
  }

  return Service;
})
